module/zfs/dmu_objset.c

```diff
@@ -1320,10 +1320,19 @@ dmu_objset_upgrade_task_cb(void *data)
 static void
 dmu_objset_upgrade(objset_t *os, dmu_objset_upgrade_cb_t cb)
 {
+	boolean_t config_held = B_FALSE;
```

Comment:
Rather than add this conditional logic to `dmu_objset_upgrade()` it looks like it might be cleaner to ASSERT that the caller has the dsl pool config log held.  That would allow you to unconditionally take the long hold as you originally intended.  Doing so should be straight forward in the `zfs_ioc_userobjspace_upgrade` and `dmu_objset_own` callers by moving down the `dsl_pool_rele` call a little bit.  As for the call in `zfs_fuid_overobjquota()` you should be able to safely take and drop the pool config lock in the upgrade case unconditionally here.

Reply:
Thank you, I'll fix this.